juju/application.py

```diff
@@ -641,48 +643,85 @@ async def refresh(
         if switch is not None and revision is not None:
             raise ValueError("switch and revision are mutually exclusive")
 
-        charms_facade = client.CharmsFacade.from_connection(self.connection)
-        resources_facade = client.ResourcesFacade.from_connection(
-            self.connection)
+        resources_facade = client.ResourcesFacade.from_connection(self.connection)
+
         app_facade = self._facade()
 
-        charmstore = self.model.charmstore
-        charmstore_entity = None
+        charm_url = self.data['charm-url']
+        parsed_url = URL.parse(charm_url)
+        charm_name = parsed_url.name
+
+        # First we need to make sure we have the resources for the charm that's
+        # coming
+
+        # Get the list of resources needed to deploy this charm
+        if Schema.CHARM_HUB.matches(parsed_url.schema):
+            # Charmhub charms
+            charmhub = self.model.charmhub
+            charm_resources = await charmhub.list_resources(charm_name)
+
+            charm_url_origin_result = await app_facade.GetCharmURLOrigin(application=charm_name)
+
+            if charm_url_origin_result.error is not None:
+                err = charm_url_origin_result.error
+                raise JujuError(f'{err.code} : {err.message}')
+            charm_url = charm_url_origin_result.url
+            origin = charm_url_origin_result.charm_origin
+
+            if channel:
+                ch = Channel.parse(channel).normalize()
+                origin.risk = ch.risk
+                origin.track = ch.track
+
+            charms_facade = client.CharmsFacade.from_connection(self.connection)
+            resolved_charm_with_channel_results = await charms_facade.ResolveCharms(resolve=[client.ResolveCharmWithChannel(
+                charm_origin=origin,
+                switch_charm=False,
+                reference=charm_url,
+            )])
+            resolved_charm = resolved_charm_with_channel_results.results[0]
+
+            if resolved_charm.error is not None:
+                err = resolved_charm.error
+                raise JujuError(f'{err.code} : {err.message}')
+            dest_origin = resolved_charm.charm_origin
+            charm_url = resolved_charm.url
 
-        if switch is not None:
-            charm_url = switch
-            if not charm_url.startswith('cs:'):
-                charm_url = 'cs:' + charm_url
         else:
-            charm_url = self.data['charm-url']
-            charm_url = charm_url.rpartition('-')[0]
-            if revision is not None:
-                charm_url = "%s-%d" % (charm_url, revision)
+            charms_facade = client.CharmsFacade.from_connection(self.connection)
+            charmstore = self.model.charmstore
+            charmstore_entity = None
+            if switch is not None:
+                charm_url = switch
+                if not charm_url.startswith('cs:'):
+                    charm_url = 'cs:' + charm_url
```

Comment:
What happens with local charms? I see a method below which is localRefresh?

I'd say it's an error if the url has no schema for refresh unless switch is used. With the cli, a url without a schema defaults to charmhub.

Reply:
Is there a test with refresh doing a switch from cs to ch? Or refresh with switch to local?

Reply:
We do have a (passing) test for local upgrade https://github.com/juju/python-libjuju/blob/bd01785ab0ee3ba2374ecc60d93bd9bc1e6c070e/tests/integration/test_application.py#L196, but I agree I should probably revise the logic for switch and add some tests for it :+1: